File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(root==NULL)
        {
            return true;
        }
        int left=maxheight(root->left);
        int right=maxheight(root->right);
        if(abs(left-right)>1)
        {
            return false;
        }

        return isBalanced(root->left)&&isBalanced(root->right);
        
    }

    int maxheight(TreeNode*root)
    {
        if(root==NULL)
        {
            return 0;
        }
        return 1+max(maxheight(root->left),maxheight(root->right));
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        int count=0;
        if(root==null)
        {
            return count;
        }
        Queue<TreeNode>q=new ArrayDeque<>();
        count=1;
        q.offer(root);
        while(!q.isEmpty())
        {
            int size=q.size();
            for(int i=0;i<size;i++)
            {
                TreeNode curr=q.poll();
                if(curr.left!=null)
                {
                    if(curr.left.val<curr.val)
                    {
                        curr.left.val=curr.val;
                    }
                    else
                    {
                        count++;
                    }
                    q.offer(curr.left);
                }
                if(curr.right!=null)
                {
                    if(curr.right.val<curr.val)
                    {
                        curr.right.val=curr.val;
                    }
                    else
                    {
                        count++;
                    }
                    q.offer(curr.right);
                }
            }
        }
        return count;

        
    }
}
